Flexible parquet struct converter

jxiang opened the pull request.

jxiang commented:
Our parquet struct schema keeps evolving. Very frequently, we have parquet files of different versions co-existing. Instead of failing the query due to schema mismatching, it is better to return null for new fields if the data file is old.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -735,7 +743,7 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
             List<Type> prestoTypeParameters = prestoType.getTypeParameters();
             List<parquet.schema.Type> fieldTypes = entryType.getFields();
             checkArgument(
-                    prestoTypeParameters.size() == fieldTypes.size(),
```
Sometimes this check helps to catch problems with the data definition in the metastore, so I think silently returning nulls if the schemas do not match is not the right way to go.

jxiang replied:
This check is good but too restrictive. It makes it hard to evolve the schema seamlessly. We return nulls and log some message below so that queries can still go on. If they pay attention to the logging, they should realize what's happened.

Even if the number matches, the schema still could be incompatible. Users still need to know what they are doing.

nezihyigitbasi commented:
can you add unit tests?

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -745,10 +753,22 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
                 converters.add(createConverter(prestoTypeParameters.get(i), columnName + "." + fieldType.getName(), fieldType, i));
             }
+            if (prestoTypeParameters.size() != fieldTypes.size()) {
+                log.info("Parquet column " + columnName + " field number mismatch, metastore has: "
```
I don't think we need the log here (especially in INFO level)

jxiang replied:
It's good to log something since we don't fail the query any more. I can make this a warning.

jxiang commented:
Sure, will add some unit tests.

jxiang commented:
Added a unit test to the patch.

jxiang commented:
Updated the patch: 1) now we support both adding and removing struct fields; 2) support adding fields at any place in the struct; 3) added more test cases for these scenarios.
However, it is not supported to change the order of existing fields in a struct. In such a case, a schema mismatch error will be thrown.

jxiang commented:
Added another patch that supports changing the order of existing fields in a struct. With these patches, now we fully support Parquet struct schema evolution. 

- The schema from the metastore is the source of truth
- Use field order from the metastore schema,
- Field in metastore schema but not in parquet schema, has null value,
- Filed not in metasotre schema but in parquet schema, is ignored.

Schema mismatch will be logged although query executes.

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -684,32 +693,40 @@ public void end()
         void afterValue();
     }
 
+    private interface BlockFieldConverter extends BlockConverter
```
Is it possible to update all BlockConverter into BlockFieldConverter? Then we just need on BlockConverter, with fieldIndex

jxiang replied:
Most of converters can share the same interface, except ParquetListEntryConverter and ParquetMapEntryConverter. The field index doesn't apply to these two entry converters.

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -684,32 +693,40 @@ public void end()
+    private interface BlockFieldConverter extends BlockConverter
+    {
+        void setFieldIndex(int fieldIndex);
+
```
no blank line here

jxiang replied:
Fixed

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -722,18 +739,18 @@ private static GroupedConverter createGroupConverter(Type prestoType, String col
         private BlockBuilder currentEntryBuilder;
 
-        public ParquetStructConverter(Type prestoType, String columnName, GroupType entryType, int fieldIndex)
+        private int[] converterFieldIndex;
```
rename to converterIndexes? what do u think?

jxiang replied:
Sure. Done.

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -722,18 +739,18 @@ private static GroupedConverter createGroupConverter(Type prestoType, String col
 
-        public ParquetStructConverter(Type prestoType, String columnName, GroupType entryType, int fieldIndex)
+        private int[] converterFieldIndex;
+        private InterleavedBlockBuilder tmpBuilder; // a tmp builder for fields out of order
```
let's not call it tmpBuilder, call it outofOrderBlockBuilder or parquetSchemaBlockBuilder, what do u think?

jxiang replied:
Ok, renamed it to outOfOrderBlockBuilder

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
+            return converters.build();
+        }
+
+        private List<BlockFieldConverter> createConvertersByName(String columnName, List<RowField> prestoTypeFields, List<parquet.schema.Type> fieldTypes)
```
createParquetNameConverters?

zhenxiao replied:
prestoRowFields, and parquetTypes

jxiang replied:
It's better to use createConvertersByName/createConvertersUseNames.
Changed prestoTypeFields to prestoRowFields. fieldTypes is fine.

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
+
+        private List<BlockFieldConverter> createConvertersByName(String columnName, List<RowField> prestoTypeFields, List<parquet.schema.Type> fieldTypes)
+        {
+            int fieldCount = fieldTypes.size();
```
prestoFieldCount

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
 
-            ImmutableList.Builder<BlockConverter> converters = ImmutableList.builder();
-            for (int i = 0; i < prestoTypeParameters.size(); i++) {
+            Map<String, ObjectPair<Integer, Type>> fieldTypeMap = new HashMap<String, ObjectPair<Integer, Type>>(prestoFieldCount);
```
how about using 2 maps? Map<String, Integer> fieldIndexes, Map<String, Type> fieldTypes

jxiang replied:
Created a private Pair class.

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
-            ImmutableList.Builder<BlockConverter> converters = ImmutableList.builder();
-            for (int i = 0; i < prestoTypeParameters.size(); i++) {
+            Map<String, ObjectPair<Integer, Type>> fieldTypeMap = new HashMap<String, ObjectPair<Integer, Type>>(prestoFieldCount);
+            for (int i = 0, n = prestoTypeFields.size(); i < n; i++) {
```
for (int i = 0; i < prestoTypeFields.size(); i++)

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
-            for (int i = 0; i < prestoTypeParameters.size(); i++) {
+            Map<String, ObjectPair<Integer, Type>> fieldTypeMap = new HashMap<String, ObjectPair<Integer, Type>>(prestoFieldCount);
+            for (int i = 0, n = prestoTypeFields.size(); i < n; i++) {
+                RowField field = prestoTypeFields.get(i);
```
try use java8 apis here

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
-                converters.add(createConverter(prestoTypeParameters.get(i), columnName + "." + fieldType.getName(), fieldType, i));
+                ObjectPair<Integer, Type> pair = fieldTypeMap.get(fieldType.getName().toLowerCase());
+                if (pair == null) {
+                    log.info("Schema mismatch, metastore schema for row column %s doesn't have field %s but parquet schema has it",
```
double logging the same thing, just keep the previous log is OK

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
+                    builder.add(ParquetDummyPrimitiveConverter.converter);
+                }
+                else {
+                    int currFieldIndex = pair.getFirst().intValue();
```
int fieldIndex or int currentFieldIndex

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
+                fieldTypeMap.put(field.getName().get().toLowerCase(), ObjectPair.create(Integer.valueOf(i), field.getType()));
+            }
+
+            boolean outOfOrder = false;
```
let's add some comments here
when reading parquet files, we need to pass converters and parquet types in parquet schema fields order
after getting the parquet values, we need to build presto blocks in metastore schema fields order

jxiang replied:
Added

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
-            this.converters = converters.build();
+            List<BlockFieldConverter> converters = builder.build();
+            if (outOfOrder) {
+                log.info("Schema mismatch, struct field order for metastore schema for row column %s is different from that of parquet schema",
```
duplicate log? just one log is OK

jxiang replied:
Not duplicate. There could be several different mismatch scenarios.

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
+            if (outOfOrder) {
+                log.info("Schema mismatch, struct field order for metastore schema for row column %s is different from that of parquet schema",
+                        columnName);
+                ImmutableList.Builder<Type> types = ImmutableList.builder();
```
add comment here: when reading parquet files, pass converters and parquet types in parquet schema fields order

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
+            checkArgument(ROW.equals(prestoType.getTypeSignature().getBase()));
+            List<parquet.schema.Type> fieldTypes = entryType.getFields();
+
+            if (useNames) {
```
use name based field lookup or index based field lookup
I am not sure whether we need another configuration to decide automatically support nested schema evolution (adding/removing/reordering fields in parquet), let's just use the existing useNames for now

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -771,15 +876,26 @@ public void start()
+                currentEntryBuilder = builder;
             }
+
+            if (tmpBuilder != null) {
```
previous 2 changes could avoid, just update currentEntryBuilder if tmpBuilder is not null

zhenxiao commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java:
```diff
@@ -790,7 +906,39 @@ public void end()
+                }
+
+                currentEntryBuilder = currentEntryBuilder.beginBlockEntry();
+                BlockFieldConverter[] converterArray = new BlockFieldConverter[converterFieldIndex.length];
```
use parquetFieldCount wherever possible, instead of converterFieldIndex.length

zhenxiao commented:
@jxiang very nice work, mostly good, I left some comments about coding style, naming, etc.
@nezihyigitbasi @dain @electrum here is some context:
Jimmy's PR is trying to support nested schema evolution for Parquet in Presto (really a common case for Uber, could be useful for other companies as well).
An example, Originally, users have parquet file with schema {a int, b struct {c int, d double}}, both parquet files and metastore are storing the this schema. After a while, users are adding/removing/reordering fields inside struct, the new schema becomes {a int, b struct {d double, c int, f float}}, all new parquet files will have the new schema, metastore is updated to the new schema, while, the old parquet files still holds the old schema  {a int, b struct {c int, d double}}, and due to the huge volume, it is not possible to convert them into new schema.

Jimmy's solution is to support querying both new parquet files and old parquet files using the new schema, if there is a field only exists in metastore, not existing in parquet files, then return null. If there is a field only exists in parquet files, not existing in metastore, ignore them.

Your comments and suggestions are appreciated.

billonahill commented:
How do queries get past the partition/table compatibility check in HiveSplitManager? Could you try applying the tests patch from PR #4939 to see if TestHiveClient integration tests pass?

jxiang commented:
#4714 supports add/remove any fields anywhere in a struct. It also supports
changing the order of fields. So it supports adding/removing fields at the
end as #4939 as well.

As to the partition/table compatibility check, #4714 doesn't change the
existing logic. Current logic as I know just check the type name, for
example, struct. It doesn't check the fields inside the struct. That's why
#4714 works for us. We have tested it in out cluster for many struct schema
evolution scenarios.

On Tue, Apr 5, 2016 at 10:21 AM, Bill Graham <notifications@github.com>
wrote:

> How do queries get past the partition/table compatibility check in
> HiveSplitManager? Could you try applying the tests patch from PR #4939
> <https://github.com/prestodb/presto/pull/4939> to see if TestHiveClient
> integration tests pass?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/prestodb/presto/pull/4714#issuecomment-205906337>
>

billonahill commented:
@jxiang let's make sure we're discussing the same use case. The use case addressed in #4939 can be seen by looking at the changes to `create-test-hive13.sql`:

1. Create a partitioned table with a schema with a nested struct.
2. Insert records.
3. Change the table schema by adding a field to the nested struct.
4. Query the nested struct.

Without changing `HiveSplitManager` a query will fail when comparing `!tableType.equals(partitionType)` because the types are just strings that includes the set of struct fields. Refering to the table in `create-test-hive13.sql` you'd get a failed comparison of `struct<s_string:string,s_double:double>` with `struct<s_string:string,s_double:double,s_boolean:boolean>`. Here's the relevant code:

https://github.com/prestodb/presto/blob/ab2c1a779e220f0363febab992f29e6ffbd4b449/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java#L250

If you apply the presto-hive/src/test/* changed from #4939 to your patch and run the integration tests you will see this failure.

jxiang commented:
@dain, I rebased the patch. Could you please take a look when you get a chance?

dain commented:
I will be out for a couple of weeks and will not have time to look at this

jxiang commented:
This patch has been running fine in our production for several months now. Could you take a look? Thanks. @cberner @dain

dain commented:
@zhenxiao now that the new parquet reader supports structs, we should add this same feature to it

zhenxiao commented:
@dain yes, we are working / stress testing it
New Parquet Reader is using nested path names to look up ColumnDescriptors, it has most schema evolution support now

markcho commented:
@jxiang Is there anything that I can do to help out with this PR?

I'm facing the same problem where we have mismatching schemas for structs due to schema evolution and it's not very feasible for us backfill the old Parquet files to match the new schemas.

I can apply this change to my fork but I think other people may find this feature useful as well.

That being said, is there a different approach to schema evolution involving Parquet files for cases similar to this, without applying this patch?

billonahill commented:
+1 to @markcho's comment. We're also in need of this patch as well. cc/ @Yaliang.

Gauravshah commented:
@zhenxiao the code still doesn't allow for struct types to evolve https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java#L743-L748

anything we can do to help in this pr ?

zhenxiao commented:
we have a rebased version here:
https://github.com/prestodb/presto/pull/6675/commits/ba5a3e4a18aba2f0ec59d99466fd27eaf4ed9ef0

Gauravshah commented:
adding the updated pull request for reference https://github.com/prestodb/presto/pull/6675

jxiang commented:
Thanks @zhenxiao, I pushed the rebased version to this branch.

jxiang commented:
@dain could you take a look when you get a chance? Thanks.

dain commented:
@jxiang yep.  I see there are two (or three) PRs related to this.  Can you help me understand which ones I should review in which order?

zhenxiao commented:
@dain this PR is the very first one, now @jxiang has all schema evolution stuff in one commit. https://github.com/prestodb/presto/pull/6675 is built on top of this

jxiang commented:
Yeah, as @zhenxiao said, this is the first one. Thanks.

nezihyigitbasi commented:
@zhenxiao @jxiang AFAIU #6675 supersedes this one. If that's correct please close this PR and then we can work on the other one.

zhenxiao commented:
continue with:
https://github.com/prestodb/presto/pull/6675
